docs/DEVELOPING.md

```diff
@@ -124,6 +127,227 @@ Take a look at the [MongoDB connector](../connectors/sources/mongo.py) for more
 
 Each connector will have their own specific behaviors and implementations. When a connector is loaded, it stays in memory, so you can come up with any strategy you want to make it more efficient. You just need to be careful not to blow memory.
 
+### Common Patterns
+
+#### Source/Client
+
+Many connector implementations will need to interface with some 3rd-party API in order to retrieve data from a remote system.
+In most instances, it is helpful to separate concerns inside your connector code into two classes.
+One class for the business logic of the connector (this class usually extends `BaseDataSource`), and one class (usually called something like `<3rd-party>Client`) for the API interactions with the 3rd-party.
+By keeping a `Client` class separate, it is much easier to read the business logic of your connector, without having to get hung up on extraneous details like  authentication, request headers, error response handling, retries, pagination, etc.
+
+#### Async
+
+The CLI uses `asyncio` and makes the assumption that all the code that has been called should not block the event loop. This makes syncs extremely fast with no memory overhead. In order to achieve this asynchronicity, source classes should use async libs for their backend.
+
+When not possible, the class should use [run_in_executor](https://docs.python.org/3/library/asyncio-eventloop.html#executing-code-in-thread-or-process-pools) and run the blocking code in another thread or process.
+
+When you send work in the background, you will have two options:
+
+- if the work is I/O-bound, the class should use threads
+- if there's some heavy CPU-bound computation (encryption work, etc), processes should be used to avoid [GIL contention](https://realpython.com/python-gil/)
+
+When building async I/O-bound connectors, make sure that you provide a way to recycle connections and that you can throttle calls to the backends. This is very important to avoid file descriptors exhaustion and hammering the backend service.
+
+### Other considerations
```

Comment:
I noticed this section is written in a style closer to a tutorial, where the author is guiding the reader through concepts (using "we") and using a specific connector as a concrete example in each case.

I'm noting it only as an observation. Maybe we can expand this into a full tutorial in the future.

Reply:
I'm not good at keeping a consistent voice from session-to-session. Getting up for coffee tends to change my style way more than it should. :) 

Do you think it would be better in a different voice until such a time when it is a "full tutorial"?

Reply:
Maybe just remove the uses of "we" and stick to "you" and "do this, do that" (instructions rather than walkthrough). Most readers probably won't notice, so not a big deal.